mino/minogrpc/server.go

```diff
@@ -0,0 +1,296 @@
+package minogrpc
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"log"
+	"math/big"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/improbable-eng/grpc-web/go/grpcweb"
+	"go.dedis.ch/fabric/mino"
+	"golang.org/x/xerrors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+)
+
+var (
+	// defaultMinConnectTimeout is the minimum amount of time we are willing to
+	// wait for a grpc connection to complete
+	defaultMinConnectTimeout = 7 * time.Second
+	// defaultContextTimeout is the amount of time we are willing to wait for a
+	// remote procedure call to finish. This value should always be higher than
+	// defaultMinConnectTimeout in order to capture http server errors.
+	defaultContextTimeout = 10 * time.Second
+)
+
+// Server represents the entity that accepts incoming requests and invoke the
+// corresponding RPCs.
+type Server struct {
+	grpcSrv *grpc.Server
+
+	cert      *tls.Certificate
+	addr      *mino.Address
+	listener  net.Listener
+	httpSrv   *http.Server
+	StartChan chan struct{}
+
+	// neighbours contains the certificate and details about known peers.
+	neighbours map[string]Peer
+
+	handlers map[string]mino.Handler
+}
+
+type ctxURIKey string
+
+const ctxKey = ctxURIKey("URI")
+
+// Roster is a set of peers that will work together
+// to execute protocols.
+type Roster []Peer
+
+// Peer is a public identity for a given node.
+type Peer struct {
+	Address     string
+	Certificate *x509.Certificate
+}
+
+// RPC represents an RPC that has been registered by a client. This struct
+// implements the mino.RPC interface, which allows clients to call an RPC that
+// will execute the provided handler.
+type RPC struct {
+	handler mino.Handler
+	srv     Server
+	uri     string
+}
+
+// Call implements the Mino.RPC interface. It calls the RPC on each provided
+// address.
+func (rpc RPC) Call(req proto.Message,
+	addrs ...*mino.Address) (<-chan proto.Message, <-chan error) {
+
+	out := make(chan proto.Message, len(addrs))
+	errs := make(chan error, len(addrs))
+
+	m, err := ptypes.MarshalAny(req)
+	if err != nil {
+		errs <- xerrors.Errorf("failed to marshal msg to any: %v", err)
+		return out, errs
+	}
+
+	sendMsg := &CallMsg{
+		Message: m,
+	}
+
+	go func() {
+		for _, addr := range addrs {
+			clientConn, err := rpc.srv.getConnection(addr.GetId())
+			if err != nil {
+				errs <- xerrors.Errorf("failed to get client conn: %v", err)
+				continue
+			}
+			cl := NewOverlayClient(clientConn)
+
+			ctx, ctxCancelFunc := context.WithTimeout(context.Background(),
+				defaultContextTimeout)
+			defer ctxCancelFunc()
+
+			header := metadata.New(map[string]string{"apiuri": rpc.uri})
```

Comment:
Why is that necessary ?

Reply:
It is used to pass the handler's URI to the gRPC call.